File: app/src/main/java/com/example/firabasefirstexperience/Fragments/ChatsFragment.java
package com.example.firabasefirstexperience.Fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.firabasefirstexperience.Adapter.UserAdapter;
import com.example.firabasefirstexperience.Model.Chat;
import com.example.firabasefirstexperience.Model.User;
import com.example.firabasefirstexperience.R;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;


public class ChatsFragment extends Fragment {

    private RecyclerView recView_displayChats;
    private UserAdapter userAdapter;
    private List<User> mUsers;

    private FirebaseUser fuser;
    private DatabaseReference reference;

    private List<String> usersList;



    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             final Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_chats, container, false);

        recView_displayChats = view.findViewById(R.id.recView_displayChats);
        recView_displayChats.setHasFixedSize(true);
        recView_displayChats.setLayoutManager(new LinearLayoutManager(getContext()));

        fuser = FirebaseAuth.getInstance().getCurrentUser();

        usersList = new ArrayList<>();

        reference = FirebaseDatabase.getInstance().getReference("Chats");

        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                usersList.clear();

                for(DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    Chat chat = snapshot.getValue(Chat.class);

                    if(chat.getSender().equals(fuser.getUid()) && (!usersList.contains(chat.getReceiver()))) {
                        usersList.add(chat.getReceiver());
                    }

                    if(chat.getReceiver().equals(fuser.getUid()) && (!usersList.contains(chat.getSender()))) {
                        usersList.add(chat.getSender());
                    }

                }

                readChats();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        return view;
    }

    private void readChats() {

        mUsers = new ArrayList<>();

        reference = FirebaseDatabase.getInstance().getReference("Users");

        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                mUsers.clear();

                for(DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    User user = snapshot.getValue(User.class);

                    for(String id : usersList) {
                        if(user.getId().equals(id)) {
                            if(mUsers.size()!= 0) {
                                for(User user1 : mUsers) {
                                    if(!user.getId().equals(user1.getId())) {
                                        mUsers.add(user);
                                        break;
                                    }
                                }
                            } else {
                                mUsers.add(user);
                            }
                        }
                    }
                }

                userAdapter = new UserAdapter(getContext(), mUsers);
                recView_displayChats.setAdapter(userAdapter);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });


    }

}

File: app/src/main/java/com/example/firabasefirstexperience/kotlin/fragment/ChatsFragment.kt
package com.example.firabasefirstexperience.kotlin.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.firabasefirstexperience.kotlin.model.Chat
import com.example.firabasefirstexperience.R
import com.example.firabasefirstexperience.kotlin.activity.SearchActivity
import com.example.firabasefirstexperience.kotlin.model.User
import com.example.firabasefirstexperience.kotlin.adapter.ChatsAdapterkt
import com.github.clans.fab.FloatingActionButton
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import java.util.ArrayList

class ChatsFragment : Fragment() {

    private var recView_displayChats: RecyclerView? = null

    private var fbUser: FirebaseUser? = null
    private var reference: DatabaseReference? = null

    private var usersList: List<String>? = null
    private var mList: List<User>? = null

    private var spinner: ProgressBar? = null


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view: View = inflater.inflate(R.layout.fragment_chats, container, false)

        spinner = view.findViewById(R.id.spinner)
        spinner?.visibility = View.VISIBLE

        recView_displayChats = view.findViewById(R.id.recView_displayChats)
        recView_displayChats?.setHasFixedSize(true)
        recView_displayChats?.layoutManager = LinearLayoutManager(view.context, RecyclerView.VERTICAL, false)




        fbUser = FirebaseAuth.getInstance().currentUser
        reference = FirebaseDatabase.getInstance().getReference("Chats")

        usersList = ArrayList()

        reference?.addValueEventListener(object : ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {

            }

            override fun onDataChange(p0: DataSnapshot) {
                (usersList as ArrayList<String>).clear()

                for (snapshot in p0.children) {

                    val chat = snapshot.getValue(Chat::class.java)
                    if (chat != null) {

                        if (chat.sender == fbUser?.uid && (!(usersList as ArrayList<String>).contains(chat.receiver))) {
                            (usersList as ArrayList<String>).add(chat.receiver!!)
                        }

                        if (chat.receiver == fbUser?.uid && (!(usersList as ArrayList<String>).contains(chat.sender))) {
                            (usersList as ArrayList<String>).add(chat.sender!!)
                        }
                    }

                }

                readChats()
                spinner?.visibility = View.GONE

            }

        })

        val fab_newPen: FloatingActionButton = view.findViewById(R.id.fab_newPen)

        fab_newPen.setOnClickListener {
            val intent = Intent(context, SearchActivity::class.java)
            startActivity(intent)
        }

        return view
    }

    private fun readChats() {
        mList = ArrayList()

        reference = FirebaseDatabase.getInstance().getReference("Users")
        reference?.addValueEventListener(object : ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {

            }

            override fun onDataChange(p0: DataSnapshot) {
                (mList as ArrayList<User>).clear()

                for (snapshot in p0.children) {
                    val user = snapshot.getValue(User::class.java)

                    for (id in (usersList as ArrayList<String>)) {
                        if (user?.id == id) {
                            if (mList?.size != 0) {
                                for (user1 in mList as ArrayList<User>) {
                                    if (user.id != user1.id) {
                                        (mList as ArrayList<User>).add(user)
                                        break
                                    }
                                }
                            } else {
                                (mList as ArrayList<User>).add(user)
                            }
                        }
                    }
                }

                val chatsAdapter = ChatsAdapterkt(mList as ArrayList<User>)
                recView_displayChats?.adapter = chatsAdapter
            }

        })
    }
}